bazel/browsers/update-script/browser-artifact.ts

```diff
@@ -0,0 +1,50 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Browser} from './browser';
+
+/** Type describing possible artifact types for browser downloads. */
+export type ArtifactType = 'driver-bin' | 'browser-bin';
+
+/** Set of known artifact extensions, including chained extensions for gzipped files. */
+const KNOWN_EXTENSIONS = new Set(['zip', 'tar.gz', 'tar.bz2', 'dmg']);
+
+/** Class describing an artifact for a browser.  */
+export class BrowserArtifact {
+  /** Extension of the artifact, derived from the download URL. */
+  extension = getArtifactExtension(this.downloadUrl);
+
+  constructor(
+    /** Instance of the browser this artifact exists for. */
+    public browser: Browser<unknown>,
+    /** Type of the artifact. */
+    public type: ArtifactType,
+    /** URL for downloading the artifact.  */
+    public downloadUrl: string,
+  ) {}
+}
+
+/**
+ * Gets the extension of a given artifact file, excluding the dot/period.
+ *
+ * Since artifact download URLs can use chained extensions as for
+ * example with `.tar.gz`, we will need to keep track of known extensions
+ * and start looking with the first dot/period we discover.
+ */
+function getArtifactExtension(filePath: string, startIdx = 0): string {
+  const separatorIdx = filePath.indexOf('.', startIdx);
+  const extension = filePath.substring(separatorIdx + 1);
+
+  if (separatorIdx === -1 || extension === '') {
+    throw new Error(`Unable to find known extension for file path: ${filePath}`);
+  } else if (KNOWN_EXTENSIONS.has(extension)) {
+    return extension;
+  }
+
+  return getArtifactExtension(filePath, separatorIdx + 1);
```

Comment:
Should we leverage `path.extname` here?

```suggestion
  function getArtifactExtension(filePath: string): string {
    const fullFilePath = filePath;
    let extension = '';

    while (path.extname(filePath) !== '') {
      const extensionChunk = path.extname(filePath);
      extension = extensionChunk + extension;
      filePath = path.basename(filePath, extensionChunk);
    }

    if (KNOWN_EXTENSIONS.has(extension)) {
      return extension;
    } 

    throw new Error(`Unable to find known extension for file path: ${fullFilePath}`);
  }
```

Reply:
Yeah, I thought using `indexOf` might be more clear, but `extname` handles the special case `a/b.c/d.tar.gz` better, so 👍 good idea!